src/core/namespace.c

```diff
@@ -560,6 +579,62 @@ static int mount_dev(BindMount *m) {
         return r;
 }
 
+static int mount_bind_dev(BindMount *m) {
+        int r;
+
+        assert(m);
+
+        /* Implements the little brother of mount_private_dev(): simply bind mounts the host's /dev into the service's
+         * /dev. This is only used when RootDirectory= is set. */
+
+        r = path_is_mount_point(bind_mount_path(m), NULL, 0);
+        if (r < 0)
+                return log_debug_errno(r, "Unable to determine whether /dev is already mounted: %m");
+        if (r > 0) /* make this a NOP if /dev is already a mount point */
+                return 0;
+
+        if (mount("/dev", bind_mount_path(m), NULL, MS_BIND|MS_REC, NULL) < 0)
+                return log_debug_errno(errno, "Failed to bind mount /dev: %m");
+
+        return 1;
+}
+
+static int mount_sysfs(BindMount *m) {
+        int r;
+
+        assert(m);
+
+        r = path_is_mount_point(bind_mount_path(m), NULL, 0);
+        if (r < 0)
+                return log_debug_errno(r, "Unable to determine whether /sys is already mounted: %m");
+        if (r > 0) /* make this a NOP if /sys is already a mount point */
+                return 0;
+
+        /* Bind mount the host's version so that we get all child mounts of it, too. */
+        if (mount("/sys", bind_mount_path(m), NULL, MS_BIND|MS_REC, NULL) < 0)
+                return log_debug_errno(errno, "Failed to mount /sys: %m");
+
+        return 1;
+}
+
+static int mount_procfs(BindMount *m) {
+        int r;
+
+        assert(m);
+
+        r = path_is_mount_point(bind_mount_path(m), NULL, 0);
+        if (r < 0)
+                return log_debug_errno(r, "Unable to determine whether /proc is already mounted: %m");
+        if (r > 0) /* make this a NOP if /proc is already a mount point */
+                return 0;
+
+        /* Mount a new instance, so that we get the one that matches our user namespace, if we are running in one */
+        if (mount("proc", bind_mount_path(m), "proc", MS_NOSUID|MS_NOEXEC|MS_NODEV, NULL) < 0)
+                return log_debug_errno(errno, "Failed to mount /proc: %m");
```

Comment:
I'd replace `"/proc"` in those error messages with `"%s", bind_mount_path(m)`...